build-logic/jvm/src/main/kotlin/gradlebuild.unittest-and-compile.gradle.kts

```diff
@@ -140,12 +139,19 @@ fun addDependencies() {
 fun addCompileAllTask() {
     tasks.register("compileAll") {
         val compileTasks = project.tasks.matching {
-            it is JavaCompile || it is GroovyCompile
+            (it is JavaCompile || it is GroovyCompile) && !it.isTestCompile()
         }
         dependsOn(compileTasks)
     }
 }
 
+fun Task.isTestCompile() =
+    listOf(
+        "compileCrossVersionTest",
```

Comment:
I wonder if we can do something better than repeating the names here. Maybe the test compilation tasks could add themselves as a dependency to compileAll instead of compileAll trying to find out all the possible test tasks. I suppose the names are also specified somewhere else, so we could maybe re-use them.

I am also wondering: Does `compileAll` currently include the Kotlin compilation? Probably it does, though via a dependency I suppose.

Reply:
The test compilation tasks are created when adding source set: https://github.com/gradle/gradle/blob/4e529e43be7272acf4a7433b1fdabfebc0494092/build-logic/integration-testing/src/main/kotlin/gradlebuild/integrationtests/shared-configuration.kt#L92 so I'd just filter it here.

Reused the `TestType` enum, and it seems to belong to `basics` subproject.

> Does compileAll currently include the Kotlin compilation? 

Yes, and it takes 32s: https://ge.gradle.org/s/taoxcvqzmlseo/timeline?details=bnooxw3jbhpue&task-path=compileTestKotlin